Adding support for Launch Templates

skang0601 opened the pull request:
…licy.

# PR o'clock

## Description

Converting from using aws_launc_configuration to aws_launch_template to support mixed_instances_policy in the worker node's autoscaling group.
Addresses https://github.com/terraform-aws-modules/terraform-aws-eks/issues/194

### Checklist

- [x] `terraform fmt` and `terraform validate` both work from the root and `examples/eks_test_fixture` directories (look in CI for an example)
- [ ] Tests for the changes have been added and passing (for bug fixes/features)
- [ ] Test results are pasted in this PR (in lieu of CI)
- [x] I've added my change to CHANGELOG.md
- [ ] Any breaking changes are highlighted above

skang0601 commented on workers.tf:
```diff
@@ -1,10 +1,80 @@
+        version            = "$Latest"
+      }
+
+      override {
```
It seems like the AWS provider devs are also waiting on Terraform 0.14 to fix everything as well...

It's currently not possible to make this overrides dynamically generated and you're required to have 2 or more of these overrides.

skang0601 commented on workers.tf:
```diff
@@ -1,10 +1,80 @@
+  }
+
+  block_device_mappings {
+    device_name = "${data.aws_ami.eks_worker.root_device_name}"
```
I'm a little worried about this. 

I could generate a list of the problem root device name by using a template file trick but I'm not sure what you all might prefer.

max-rocket-internet replied:
I don't think we need this?

skang0601 replied:
The API will complain if you do not specify this. This field is required on launch templates.

max-rocket-internet commented:
Hey @skang0601,

Super excited to have Launch Template support!

Do you think we could implement in a way that does not replace the existing  arrangement with `aws_launch_configuration`? Like could we choose on a per worker group basis whether a group uses a `aws_launch_configuration` or `aws_launch_template`?

Reason being this is quite a large change with a lot of extra options and I think it might be nice to have it optional for a while so people can test it out.

What do you think? And others?!?

skang0601 commented:
@max-rocket-internet I think that's a valid ask. I took the naive approach and just created a separate input for worker groups to be created using the launch template.
Either in the future, this can probably be cleaned up further by switching to launch templates entirely or when some of the newer HCL features finally land on a Terraform release.

antonbabenko commented on examples/eks_test_fixture/main.tf:
```diff
@@ -39,18 +39,27 @@ locals {
       additional_security_group_ids = "${aws_security_group.worker_group_mgmt_one.id},${aws_security_group.worker_group_mgmt_two.id}"
+      override_instance_type        = "t3.small"
+      asg_desired_capacity          = "2"
+      override_instance_type        = "t3.small"
```
Same as line 58.

skang0601 replied:
Fixed

skang0601 commented:
Made some additional updates.
I moved the mixed_instances_policy param from a global config to per worker group configuration and updated the changelog

runningman84 commented:
is there any eta for release of this feature?

max-rocket-internet commented on variables.tf:
```diff
@@ -60,21 +60,36 @@ variable "vpc_id" {
   description = "VPC where the cluster and workers will be deployed."
 }
 
-variable "worker_groups" {
```
Like I said before, can we make this PR backwards compatible? i.e leave this variable and all related stuff in place so that a user can use launch templates if they want but without destroying existing resources.

skang0601 replied:
Sorry, misinterpreted your previous ask there.
Will update PR

max-rocket-internet commented on workers.tf:
```diff
@@ -1,20 +1,47 @@
-resource "aws_autoscaling_group" "workers" {
```
This will cause the existing ASG to be destroyed.

skang0601 replied:
Fixed

max-rocket-internet commented on workers.tf:
```diff
@@ -24,33 +51,117 @@ resource "aws_autoscaling_group" "workers" {
   }
 }
 
+resource "aws_iam_instance_profile" "launch_configuration_workers" {
```
Do we need a second instance profile? Can't both LC and LT type ASGs use the same one?

max-rocket-internet replied:
Ahh, yes, I think it needs to stay.

skang0601 replied:
Yep, especially since this change will support launching with both types.

max-rocket-internet commented on README.md:
```diff
@@ -119,8 +124,6 @@ MIT Licensed. See [LICENSE](https://github.com/terraform-aws-modules/terraform-a
 | vpc_id | VPC where the cluster and workers will be deployed. | string | - | yes |
 | worker_additional_security_group_ids | A list of additional security group ids to attach to worker instances | list | `<list>` | no |
-| worker_group_count | The number of maps contained within the worker_groups list. | string | `1` | no |
-| worker_groups | A list of maps defining worker group configurations. See workers_group_defaults for valid keys. | list | `<list>` | no |
```
To make the ever growing list of variables easier to read, perhaps let's keep them grouped like this?

```
worker_groups
worker_group_count
worker_groups_launch_template
worker_group_launch_template_count
```

skang0601 replied:
Fixed

max-rocket-internet commented:
Maybe we just create `workers_launch_template.tf` to keep everything clear and separate?

max-rocket-internet commented on aws_auth.tf:
```diff
@@ -20,7 +20,16 @@ resource "null_resource" "update_config_map_aws_auth" {
 
 data "aws_caller_identity" "current" {}
 
-data "template_file" "worker_role_arns" {
```
Can we leave this resource named the same? And call the second one `worker_launch_template_role_arns` to keep it consistent?

max-rocket-internet commented on examples/eks_test_fixture/main.tf:
```diff
@@ -39,17 +39,26 @@ locals {
   # ]
 
-  worker_groups = [
+  launch_configuration_worker_groups = [
```
Can we keep this naming consistent also? i.e. keep this as `worker_groups` and name the new one `worker_groups_launch_template`.

max-rocket-internet commented on examples/eks_test_fixture/main.tf:
```diff
@@ -127,15 +136,17 @@ module "vpc" {
+  vpc_id                                  = "${module.vpc.vpc_id}"
+  worker_groups                          = "${local.launch_configuration_worker_groups}"
+  worker_groups_launch_template           = "${local.launch_template_worker_groups}"
+  worker_group_count = "1"
```
`terraform fmt`

max-rocket-internet commented on local.tf:
```diff
@@ -10,31 +10,39 @@ locals {
   kubeconfig_name          = "${var.kubeconfig_name == "" ? "eks_${var.cluster_name}" : var.kubeconfig_name}"
 
   workers_group_defaults_defaults = {
```
This is a little confusing as now we have the defaults for both worker group types mixed together.

How would someone reading this now what parameter is valid for each type of work group?

How about we leave `workers_group_defaults_defaults` as it is and create `workers_group_launch_template_defaults_defaults`? Does that make sense and work?

skang0601 replied:
I thought about that but it would result in some duplication if we wanna make it very separate if you're fine with that.
The other approach is to just merge the two defaults for the launch_template but I think that's as confusing as mixing the two.

max-rocket-internet replied:
Hmmm I see your point about duplication.

At this point I really think the separation is important since launch template is a such a new feature and I don't want it to interfere with the current ASG/LC arrangement until everyone is confident in the ASG/LT approach.

So how about we have `workers_group_defaults_defaults` and `workers_group_launch_template_defaults_defaults` but the latter only contains the special values for LTs? e.g. `spot_max_price`, `spot_instance_pools` etc.

I think this will be clear and also won't change the existing code at all?

skang0601 replied:
I had a chance to think about this for a bit. I ended up just duplicating the defaults along.

max-rocket-internet commented on workers_launch_template.tf:
```diff
@@ -0,0 +1,103 @@
+}
+
+resource "aws_iam_instance_profile" "workers_launch_template" {
+  name_prefix = "${aws_eks_cluster.this.name}"
```
Will this name always be unique if someone is using both types of worker groups?

skang0601 replied:
I'm not completely sure what you mean by this.

max-rocket-internet replied:
No worries, I think what you have now is good 🙂

max-rocket-internet commented:
Also please rebase 🙂

max-rocket-internet commented on README.md:
```diff
@@ -106,56 +106,61 @@ Many thanks to [the contributors listed here](https://github.com/terraform-aws-m
+| map_roles | Additional IAM roles to add to the aws-auth configmap. See examples/eks_test_fixture/variables.tf for example format. | list | `<list>` | no |
+| map_roles_count | The count of roles in the map_roles list. | string | `0` | no |
+| map_users | Additional IAM users to add to the aws-auth configmap. See examples/eks_test_fixture/variables.tf for example format. | list | `<list>` | no |
+| map_users_count | The count of roles in the map_users list. | string | `0` | no |
```
I don't think any of this needs to be changed so I think you are running on older version of terraform-docs that is removing the `\_`? I'm using version `0.5.0`. Can you install that version and then fix these by running the commit hook again or fixing it manually?

skang0601 replied:
Oh wow sorry about that, I didn't realize that it produced this...
Let me update.

max-rocket-internet commented:
Thanks @skang0601! I did some testing in my envs and it looks nice. After you fix the readme changes I think we can merge 🎉

skang0601 commented:
> Thanks @skang0601! I did some testing in my envs and it looks nice. After you fix the readme changes I think we can merge 🎉

I believe I've resolved the above issues. @max-rocket-internet

zytek commented on examples/eks_test_fixture/main.tf:
```diff
@@ -41,15 +41,24 @@ locals {
+      additional_security_group_ids            = "${aws_security_group.worker_group_mgmt_one.id},${aws_security_group.worker_group_mgmt_two.id}"
+      override_instance_type                   = "t3.small"
+      asg_desired_capacity                     = "2"
+      spot_instance_pools                      = 10
```
Does not really make sense to request 2 instances across 10 spot pools. Such configuration is confusing, especially when put in examples. I would go with something along `asg_desired_capacity = 4` and `spot_instance_pools = 2` - 2x2 instance types = four instances across two spot pools (types).

max-rocket-internet replied:
I'm gonna merge this now as it's been open and on-going for a while now. The LT/ASG arrangement is also completely optional for now and not used by default. By merging this people can at least have a play with it for now.

@zytek perhaps you could make a PR with the changes suggested here?

zytek replied:
Awesome. I'll play with it and can make a PR.
Could you push new version to terraform repository ?

max-rocket-internet replied:
> Could you push new version to terraform repository ?

Sure but in a couple days. This repo has sucked up enough of my time for today. You can always point your module source directly at github until a new release is created 🙂

RothAndrew replied:
FYI:

```
module "eks" {
  source = "git::https://github.com/terraform-aws-modules/terraform-aws-eks.git?ref=whatever_ref_you_want_goes_here"
}
